envoy/http/header_map.h

```diff
@@ -801,6 +803,25 @@ class HeaderMatcher {
 
 using HeaderMatcherSharedPtr = std::shared_ptr<HeaderMatcher>;
 
+/**
+ * Base class for both tunnel response headers and trailers.
+ */
+class TunnelResponseHeadersOrTrailers : public StreamInfo::FilterState::Object {
+public:
+  ProtobufTypes::MessagePtr serializeAsProto() const override {
```

Comment:
I think it would be better to keep this private and move the base implementation to `source/common/http`. Take a look at `StringAccessor` and `StringAccessorImpl`.

Reply:
moved implementation, thanks!